### PR TITLE
Fix gitoliteHost validation

### PR DESCRIPTION
**Pull request by willdollman**
## Test plan

- [x] Update unit tests

### CODE REVIEW

**Review by evict — APPROVED** [review]

File: internal/security/security.go [hunk 1]
```diff
@@ -29,6 +31,23 @@ func ValidateRemoteAddr(raddr string) bool {
...
+	// Check if the string contains a username (e.g. git@example.com); if so validate username
+	fragments := strings.Split(raddr, "@")
+	// raddr contains more than one `@`
+	if len(fragments) > 2 {
+		return false
+	}
+	// raddr contains exactly one `@`
+	if len(fragments) == 2 {
+		user := fragments[0]
+		if match, err := regexp.MatchString("^[a-zA-Z0-9]+$", user); !match || err != nil {
```
Comment (ryanslade) [inline internal/security/security.go:43]: You can move this out as a global variable and use the `lazyregexp` package so we only need to build the regular expression once. 

Some examples:
https://sourcegraph.sourcegraph.com/search?q=context%3Aglobal+repo%3A%5Egithub%5C.com%2Fsourcegraph%2Fsourcegraph%24+lazyregexp.New&patternType=standard&sm=1&groupBy=path

**Review by evict — APPROVED** [review]

File: cmd/gitserver/server/list_gitolite_test.go [hunk 2]
```diff
@@ -25,6 +25,40 @@ func Test_Gitolite_listRepos(t *testing.T) {
...
+				{
+					Host:   "git@sourcegraph.com",
+					Prefix: "sourcegraph.com/",
+				},
+			},
+			gitoliteHost:    "git@sourcegraph.com",
+			expResponseCode: 200,
+			expResponseBody: `[{"Name":"myrepo","URL":"git@sourcegraph.com:myrepo"}]` + "\n",
+		},
+		// Invalid gitoliteHost (fails ValidateRemoteAddr)
```
[the remark this answers is not in the document]
Comment (willdollman) [reply cmd/gitserver/server/list_gitolite_test.go]: Great catch, this crossed my mind but didn't follow up. Fixed!

File: cmd/gitserver/server/list_gitolite.go [hunk 3]
```diff
@@ -30,11 +30,15 @@ func (g gitoliteFetcher) listRepos(ctx context.Context, gitoliteHost string, w h
...
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
-		}
+	// 🚨 SECURITY: If gitoliteHost is a non-empty string that fails hostname validation, return an error
+	if gitoliteHost != "" && !security.ValidateRemoteAddr(gitoliteHost) {
+		http.Error(w, "invalid hostname", http.StatusInternalServerError)
+		return
+	}
+
+	if repos, err = g.client.ListRepos(ctx, gitoliteHost); err != nil {
```
Comment (ryanslade) [inline cmd/gitserver/server/list_gitolite.go:39]: The above check can allow an empty  value for `gitoliteHost`, is that ok here?

Comment (ryanslade) [reply cmd/gitserver/server/list_gitolite.go:39]: Nevermind, I see a test case for that below 😄

**Review by ryanslade — APPROVED** [review]

**Pull request merged**